runtime-modules/content-directory/src/lib.rs

```diff
@@ -1198,37 +1268,62 @@ decl_module! {
             property_values: BTreeMap<PropertyId, PropertyValue<T>>
         ) -> dispatch::Result {
 
+            // Retrieve Class, Entity and ensure given have access to the Entity under given entity_id
             let (class, entity, _) = Self::ensure_class_entity_and_access_level(origin, entity_id, &actor)?;
 
+            // Ensure Schema under given id is not added to given Entity yet
             entity.ensure_schema_id_is_not_added(schema_id)?;
 
-            let schema = class.ensure_schema_exists(schema_id)?;
+            // Ensure provided property_values are not added to the Entity values map yet
+            entity.ensure_property_values_are_not_added(&property_values)?;
+
+            // Ensure Class Schema under given index exist, return corresponding Schema
+            let schema = class.ensure_schema_exists(schema_id)?.to_owned();
 
+            // Ensure provided schema can be added to the Entity
             schema.ensure_is_active()?;
 
-            let entity_values = entity.get_values_ref();
+            // Ensure all provided property values are for properties in the given schema
+            schema.ensure_has_properties(&property_values)?;
```

Comment:
That's not a case, because `ValuesForExistingProperties` are not computed from `Schema` `property_ids`. They are computed from `PropertyValue`'s  and their corresponding `Class` `Property`'s  https://github.com/iorveth/substrate-runtime-joystream/blob/update_entity_permissions_extrinsic/runtime-modules/content-directory/src/helpers.rs#L51, which ids can easily be totally different from `Schema` `property ids` (non required properties)

Reply:
As discussed internally, while `Schema::ensure_has_properties` is not the right check itself, `ValuesForExistingProperties` computation can be made safe, and this is done by lifting it up earlier and returning `Err` on bad input.